### ISSUE TITLE
Allow static assignments within the DHCP pool

### ISSUE THREAD

**Issue by StopTrain** [opnsense/core#35]
I know that it was a common practice for ISC DHCP being used within other distributions to not allow static assignments within the DHCP pool, however, it defeats the purpose of allow said static assignments.

would like to request the removal of the restriction in the GUI and also update the wording in the GUI with the warning that it can cause issues blah blah blah, but allow us to assign static DHCP leases within the pool.

Thank you!

**Comment by pdf** [comment]
You really don't want to do this.  Assign your static addresses from one part of the subnet, and assign your dynamic addresses from another part of the subnet.  There's no other way to do it reliably, and if OPNsense is stopping you from shooting yourself in the foot, that sounds fine to me.

**Comment by StopTrain** [comment]
And why is this 'shooting myself in the foot?'  

I've been doing this for years without issue.

**Comment by pdf** [comment]
Because at some stage the DHCP server will allocate the same address to a dynamic client that you have assigned to a fixed client, and then both devices will have broken networking.  ISC-DHCP doesn't do any checks on this because it assumes the administrator isn't doing something stupid.

**Comment by StopTrain** [comment]
Is this something you've seen for yourself?  I've been using ISC-DHCP for a very long time and never experienced this.

**Comment by pdf** [comment]
Yes, I have seen this in the wild.  You've just been lucky (and probably you aren't running networks with a lot of hosts, so that reduces the chance of collisions, but does not eliminate it).

**Comment by StopTrain** [comment]
I've run ISC-DHCP on quite a large number of subnets, never seen it happen.

**Comment by StopTrain** [comment]
I guess the bigger question is, if this IS a bigger issue with ISC-DHCP, is there an issue with switching that portion over to use dnsmasq instead?

**Comment by pdf** [comment]
Why won't you just do the right thing?  If you doubt this is an issue, go and do your own research.  Switching DHCP providers is non-trivial, and if the only reason for promoting the change is to support a configuration that's easily catered for by configuring your network properly, I don't see anyone jumping to do the work when there are more important things to get done.

**Comment by StopTrain** [comment]
There's no reason to be a prick about it

**Comment by pdf** [comment]
Sorry if what I said came across that way, those are just the facts.

**Comment by fichtner** [comment]
Guys, keep it friendly. There is no reason to not sell rope to anyone unless you'll tell them to go hang themselves with it. Less restrictions mean less code clutter, so if anybody feels like reopening this that would be fine... :)

**Comment by pdf** [comment]
@fichtner I wasn't trying to be antagonistic, and I'll apologize again it appeared that way.

I'm not sure if that's a great direction to take though - obviously there are people who may not be aware that it's bad practice, or why, and they'll be the ones who struggle to troubleshoot when it bites them.  The code's already written, and the maintenance overhead for this sort of validation should be very minimal, I'd definitely vote for keeping it as is.

**Comment by fichtner** [comment]
@pdf The only thing I vote for is for everyone to bring their interests to the table and participate in the discussion so that there can be opportunity to learn something. And maybe afterwards we'll fix a bug or two or change the way we approach a problem completely. The outcome shouldn't really matter (we'll do something sensible in any case) and there should be room for the benefit of the doubt. :)

**Comment by StopTrain** [comment]
@pdf I took it that way, but no apology necessary.  The issue was that my thinking was completely unclear.  I actually did some in depth research and confirmed that ISC themselves do not recommend adding static leases with the DHCP pool, however, this is changing with the new Kea software they are building: http://kea.isc.org/

Might want to keep an eye on this.  This will allow for proper static assignments within the actual DHCP pool.
